File: Longest_String_Chain.cpp
class Solution {
public:
    int longestStrChain(vector<string>& words) {
        std::unordered_map<std::string, int> chains;  // Stores the max chain length for each word
        std::vector<std::string> sortedWords = words;
        std::sort(sortedWords.begin(), sortedWords.end(), [](const std::string& a, const std::string& b) {
            return a.length() < b.length();  // Sort words by length
        });

        for (const std::string& word : sortedWords) {
            chains[word] = 1;  // Initialize the chain length for the current word

            for (int i = 0; i < word.length(); i++) {
                std::string pred = word.substr(0, i) + word.substr(i + 1);  // Generate predecessor by removing one character
                if (chains.find(pred) != chains.end()) {
                    chains[word] = std::max(chains[word], chains[pred] + 1);
                }
            }
        }

        int maxChainLength = 0;
        for (const auto& entry : chains) {
            maxChainLength = std::max(maxChainLength, entry.second);
        }

        return maxChainLength;        
    }
};

File: Longest_String_Chain.java
class Solution {
    public int longestStrChain(String[] words) {
        Map<String, Integer> chains = new HashMap<>();  // Stores the max chain length for each word
        String[] sortedWords = Arrays.copyOf(words, words.length);
        Arrays.sort(sortedWords, (a, b) -> a.length() - b.length());  // Sort words by length

        for (String word : sortedWords) {
            chains.put(word, 1);  // Initialize the chain length for the current word

            for (int i = 0; i < word.length(); i++) {
                String pred = word.substring(0, i) + word.substring(i + 1);  // Generate predecessor by removing one character
                if (chains.containsKey(pred)) {
                    chains.put(word, Math.max(chains.getOrDefault(word, 0), chains.get(pred) + 1));
                }
            }
        }

        int maxChainLength = chains.values().stream().mapToInt(Integer::intValue).max().orElse(0);
        return maxChainLength;        
    }
}
